### ISSUE TITLE
$class::add_extension() and missing addWrapperMethods

### ISSUE THREAD

**Issue by axllent** [silverstripe/silverstripe-framework#8630]
I upgraded a site from 4.2.1 to 4.2.2 and experiencing an issue that I am having great difficulties explaining (but I'll do my best).

What I am experiencing is some inconsistencies when extending classes statically. What appears to he happening is that classes that are extended with `$class::add_extension()` are losing (or aren't being assigned) their `addWrapperMethod`s (ie: from `Dataobject::defineMethods()`);

Although it doesn't appear to affect all DataObjects, it is as simple as `MyDataObject::add_extension(<EmptyDataExtensionClass>);` and one looses all access to all has_many, many_many etc associations. Extending the same DataObject with the same Extension via yaml works however.

As I said, not all DataObjects are affected, so I'm guessing a race condition or something, I really don't know.

This issue appears only to affect SilverStripe 4.2.2 and hasn't existed since 4.0-alpha.

I hope someone can shed some light on this please?

**Comment by robbieaverill** [comment]
Possibly related to #8613

**Comment by chillu** [comment]
Closing in favour of other ticket. @axllent if that PR doesn't address the issue in your context, please reopen.

**Comment by axllent** [comment]
Would you mind referencing that PR/Ticket @chillu, or is it the one mentioned earlier (#8613)?

**Comment by tractorcow** [comment]
I'm going to re-open this ticket, since I don't believe that these issues are the same. #8613 relates to remove_extension() and class hierarchy and this with add_extension() and extra methods.

**Comment by maxime-rainville** [comment]
I'm setting the impact to high here because this looks like a regression in a patch release.

I've had a go at reproducing this, but didn't manage to however. I tried applying an extension to a Dummy DataObject in my `_config.php` to see if I would loose extensions applied via my YML config. I tried renaming the classes to see if it would trigger some race condition.

Can anyone find a systematic way of triggering the described behaviour?

**Comment by axllent** [comment]
Hi @maxime-rainville - this is a hard one to reproduce without sending you down a rabbit hole, however it is related to a specific module of mine ([ftsearch](https://github.com/axllent/silverstripe-ftsearch)). I'll try summarise....

The module allows you to configure any DataObject to be included in a search, and automatically extends (based on whether it is Versioned or not) relevant relationship models to trigger events (ie: relating to search reindexing). This extending is done statically via [this function](https://github.com/axllent/silverstripe-ftsearch/blob/master/src/SearchEngine.php#L119).

Anyway, when this module is enabled I loose controller access to (for example) some page's has_many objects, so templates started showing up with missing data.

I apologise again but there is no easy "just do this and you get the issue" scenario. If you prefer to discuss it further outside of the issue tracker then just let me know please (and where, slack?).

**Comment by maxime-rainville** [comment]
I had another look with `ftsearch` installed. Didn't manage to replicate the issue.

Your `attachFTSearchListener()` method does some unusual things. I'm not sure if this is something we should or shouldn't support. Your method is accessing and editing a lot of configuration and relations definitions and extensions from a `_config.php`, which may be triggering the race condition you were talking about.

I'm not quite sure what could have change in 4.2.2 to break this.

**Comment by axllent** [comment]
Thank you for looking into that @maxime-rainville. I am aware it is complicated - the idea behind it was that the `attachFTSearchListener()` would recursively extend a series of dataobjects (those to be indexed), and automatically extending associated dataobjects so that "triggers" are set for when a change is detected. The only way I saw possible was extending them via the static `DO::add_extension()`, which as I pointed out has always worked as expected.

So with that concept in mind, how would you suggest I approach the solution in a "usual manner"? I'm completely open to suggestions, I don't mind rewriting the entire thing, in a manner that doesn't require one manually having to extend each and every object (and all associated DOs) manually and individually.

Given you can't replicate the problem it makes it difficult to find out why it is not working. I could (if you wanted) provide you with the actual code + sspak where this issue does happen.

**Comment by kinglozzer** [comment]
> I don't mind rewriting the entire thing, in a manner that doesn't require one manually having to extend each and every object (and all associated DOs) manually and individually.

@axllent Have you considered taking the same approach as the `silverstripe-fulltextsearch` module? That module wraps _all_ database manipulations in [a “proxy”](https://github.com/silverstripe/silverstripe-fulltextsearch/blob/master/src/Search/Extensions/ProxyDBExtension.php) that then [records the classes being modified](https://github.com/silverstripe/silverstripe-fulltextsearch/blob/master/src/Search/Updaters/SearchUpdater.php#L68-L124) and triggers an index update if the table being modified [is one that should be indexed](https://github.com/silverstripe/silverstripe-fulltextsearch/blob/master/src/Search/Updaters/SearchUpdater.php#L144-L146). That approach should save dynamically adding extensions on each request, which is quite an expensive operation

**Comment by axllent** [comment]
@kinglozzer That is a very interesting approach, thanks for the info! Agreed, I am fully aware of the cost of statically adding extensions on each request, although to be fair the tests I did barely made any difference once I "streamlined" the attachment process. Trust me, I did a lot of testing to get this "right" when writing the module as I couldn't see another way at the time, and was very aware of the potential to slow things down.

So doing it through a proxy will require a compete rewrite, and I'm not even sure it would work.

1. The the proxy detects data manipulation of a table
2. Works out the class owning that table
3. Detects whether it or any of its related dataobjects are indexed in the fulltext
4. Re-indexes that object

Somewhere in there it still has to pass on the DB request as originally intended, though I'd assume this is probably easy (I haven't read through all the code yet).

So in theory straight forward, but in practice it still has to deal with versioned and nonversioned objects, inserted, updated, archived and deleted "states", and I can sense a world of pain if trying to work that out from the database upwards, as we'd be dealing with all tables presumably, including relationship tables, versioned, `_Live` etc. It is probably also going to be somewhat expensive I would think, though only gets called with database data manipulation. I'll investigate this option further though as I think it has potential. Thanks again.

**Comment by maxime-rainville** [comment]
I'll close this issue for now, as I think this is more an issue with the module than with framework. 

If you find a consistent way of replicating the issue, I'm happy to reopen and have another look.

**Comment by axllent** [comment]
FYi: I have managed to solve the issue (or maybe it's a work-around, I'm not sure). Calling a singleton prior to adding the extension ([see here](https://github.com/axllent/silverstripe-ftsearch/blob/6daf9778c8ecfcea6bc6b64c98c64c86c762a2ef/src/SearchEngine.php#L127-L135)) causes these issues (where I am experiencing it). I also tried with `DataObject::singleton()` which had the same problem. If I [add the extension first](https://github.com/axllent/silverstripe-ftsearch/blob/7e79caec50913f9bb773b3c2e2f01f0b343e3337/src/SearchEngine.php#L159-L167) and then call the singleton the problem is solved.

Maybe this means more to some of you than it does to me?! Thanks again for looking into it.
